Source/WebKit/UIProcess/API/glib/WebKitImage.cpp

```diff
@@ -0,0 +1,333 @@
+        break;
+    case PROP_HEIGHT:
+        image->priv->height = g_value_get_int(value);
+        break;
+    case PROP_STRIDE:
+        image->priv->stride = g_value_get_uint(value);
+        break;
+    case PROP_PIXELS:
+        image->priv->pixels = static_cast<GBytes*>(g_value_get_boxed(value));
+        break;
+    default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID(object, propId, paramSpec);
+    }
+}
+
+static void webkitImageGetProperty(GObject* object, guint propId, GValue* value, GParamSpec* paramSpec)
+{
+    auto* image = WEBKIT_IMAGE(object);
+
+    switch (propId) {
+    case PROP_WIDTH:
+        g_value_set_int(value, webkit_image_get_width(image));
+        break;
+    case PROP_HEIGHT:
+        g_value_set_int(value, webkit_image_get_height(image));
+        break;
+    case PROP_STRIDE:
+        g_value_set_uint(value, webkit_image_get_stride(image));
+        break;
+    case PROP_PIXELS:
+        g_value_set_boxed(value, webkit_image_get_pixels(image));
+        break;
+    default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID(object, propId, paramSpec);
+    }
+}
+
+static void webkit_image_class_init(WebKitImageClass* imageClass)
+{
+    GObjectClass* objectClass = G_OBJECT_CLASS(imageClass);
+    objectClass->set_property = webkitImageSetProperty;
+    objectClass->get_property = webkitImageGetProperty;
+
+    /**
+     * WebKitImage:width:
+     *
+     * The image width
+     */
+    sObjProperties[PROP_WIDTH] =
+    g_param_spec_int(
+        "width",
+        nullptr, nullptr,
+        0, G_MAXINT, 0,
+        static_cast<GParamFlags>(WEBKIT_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY));
+
+    /**
+     * WebKitImage:height:
+     *
+     * The image height
+     */
+    sObjProperties[PROP_HEIGHT] =
+    g_param_spec_int(
+        "height",
+        nullptr, nullptr,
+        0, G_MAXINT, 0,
+        static_cast<GParamFlags>(WEBKIT_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY));
+
+    /**
+     * WebKitImage:stride:
+     *
+     * The image stride
+     */
+    sObjProperties[PROP_STRIDE] =
+    g_param_spec_uint(
+        "stride",
+        nullptr, nullptr,
+        0, G_MAXUINT, 0,
+        static_cast<GParamFlags>(WEBKIT_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY));
+
+    /**
+     * WebKitImage:pixels:
+     *
+     * The image pixels
+     */
+    sObjProperties[PROP_PIXELS] =
+    g_param_spec_boxed(
+        "pixels",
+        nullptr, nullptr,
+        G_TYPE_BYTES,
+        static_cast<GParamFlags>(WEBKIT_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY));
+
+    g_object_class_install_properties(objectClass, N_PROPERTIES, sObjProperties.data());
+}
+
+/**
+ * webkit_image_new:
+ * @width: the image width
+ * @height: the image height
+ * @stride: the image stride
+ * @pixels: the image pixels
+ *
+ * Create a new #WebKitImage for the given parameters
+ *
+ * Returns: (transfer full): a #WebKitImage
+ */
+WebKitImage* webkit_image_new(int width, int height, guint stride, GBytes* pixels)
```

Comment:
I would leave this as private. Sure, one can always go through `g_object_new()`, but at the moment we only expect WebKit to create instances of this type. That's surely the case for favicons, and I suspect it will be the same for WebExtensions (cc/ @ItsJamie9494)

Reply:
Yes, with WebExtensions it would only be created internally.

Reply:
Thanks, I've moved this into the private header